packages/compiler/core/types.ts

```diff
@@ -414,163 +409,263 @@ export type Expression =
-  kind: SyntaxKind.IntersectionExpression;
-  options: Expression[];
+  readonly kind: SyntaxKind.IntersectionExpression;
+  readonly options: readonly Expression[];
 }
 
 export interface TypeReferenceNode extends BaseNode {
-  kind: SyntaxKind.TypeReference;
-  target: MemberExpressionNode | IdentifierNode;
-  arguments: Expression[];
+  readonly kind: SyntaxKind.TypeReference;
+  readonly target: MemberExpressionNode | IdentifierNode;
+  readonly arguments: readonly Expression[];
 }
 
 export interface TemplateParameterDeclarationNode extends BaseNode {
-  kind: SyntaxKind.TemplateParameterDeclaration;
-  id: IdentifierNode;
+  readonly kind: SyntaxKind.TemplateParameterDeclaration;
+  readonly id: IdentifierNode;
   symbol?: TypeSymbol;
 }
 
+//#region BoundSyntaxTree
+// prettier-ignore
+export type BindNode<T extends Node> = 
+  T extends CadlScriptNode          ? BoundCadlScriptNode : 
+  T extends ImportStatementNode     ? BoundImportStatementNode : 
+  T extends ModelStatementNode      ? BoundModelStatementNode : 
+  T extends NamespaceStatementNode  ? BoundNamespaceStatementNode : 
+  T extends InterfaceStatementNode  ? BoundInterfaceStatementNode : 
+  T extends UnionStatementNode      ? BoundUnionStatementNode : 
+  T extends UsingStatementNode      ? BoundUsingStatementNode : 
+  T extends EnumStatementNode       ? BoundEnumStatementNode : 
+  T extends AliasStatementNode      ? BoundAliasStatementNode : 
+  T extends OperationStatementNode  ? BoundOperationStatementNode : 
+  T extends EmptyStatementNode      ? BoundEmptyStatementNode : 
+  T extends InvalidStatementNode    ? BoundInvalidStatementNode : 
+  T extends DecoratorExpressionNode ? BoundDecoratorExpressionNode : 
+  T;
+
+export type BoundStatement =
+  | BoundImportStatementNode
+  | BoundModelStatementNode
+  | BoundNamespaceStatementNode
+  | BoundInterfaceStatementNode
+  | BoundUnionStatementNode
+  | BoundUsingStatementNode
+  | BoundEnumStatementNode
+  | BoundAliasStatementNode
+  | BoundOperationStatementNode
+  | BoundEmptyStatementNode
+  | BoundInvalidStatementNode;
+
+export type BoundDeclaration =
+  | BoundModelStatementNode
+  | BoundInterfaceStatementNode
+  | BoundUnionStatementNode
+  | BoundNamespaceStatementNode
+  | BoundOperationStatementNode
+  | TemplateParameterDeclarationNode
+  | BoundEnumStatementNode
+  | BoundAliasStatementNode;
+
+export type BoundScopeNode =
+  | BoundNamespaceStatementNode
+  | BoundModelStatementNode
+  | BoundInterfaceStatementNode
+  | BoundAliasStatementNode
+  | BoundCadlScriptNode;
+
+export type BoundNode =
+  | BoundCadlScriptNode
+  | TemplateParameterDeclarationNode
+  | ModelPropertyNode
+  | UnionVariantNode
+  | BoundOperationStatementNode
+  | NamedImportNode
+  | EnumMemberNode
+  | ModelSpreadPropertyNode
+  | BoundDecoratorExpressionNode
+  | DirectiveExpressionNode
+  | BoundStatement
+  | Expression;
+
+export type BindBaseNode<T extends BaseNode> = T & {
```

Comment:
```suggestion
export type BoundBaseNode<T extends BaseNode> = T & {
```

And also all BindXxx to BoundXxx (slightly different than prior round of binded to bound )